python/tests/test_bindings.py

```diff
@@ -0,0 +1,94 @@
+import unittest
+import hyperonpy as hp
+from hyperon import *
+from copy import deepcopy
+
+
+class BindingsTest(unittest.TestCase):
+
+    def setUp(self):
+        self.emptyBindings = Bindings()
+
+        self.bindings = Bindings()
+        self.bindings.add_var_bindings("a", SymbolAtom(hp.atom_sym("b")))
+        self.bindings.add_var_bindings("x", SymbolAtom(hp.atom_sym("y")))
+
+
+    def tearDown(self) -> None:
+        pass
+
+    def test_bindings_match_display(self):
+        pass
+        '''
+        commented till sort inside bindings become stable
+        bindings = hp.bindings_new()
+
+        # uncomment this and assert on line 22 fails
+        #self.assertEqual(hp.bindings_to_str(bindings), "{  }")
+        hp.bindings_add_var_bindings(bindings, "a", hp.atom_sym("b"))
+        hp.bindings_add_var_bindings(bindings, "x", hp.atom_sym("y"))
+
+        bindings_as_str = hp.bindings_to_str(bindings)
+
+        # sometimes fails
+        self.assertEqual(bindings_as_str, "{ $x = y, $a = b }")
+
+        hp.bindings_free(bindings)
+        '''
+
+    def test_bindings_equality_and_clone(self):
+
+        with deepcopy(self.emptyBindings) as cloned_empty:
+            self.assertEqual(self.emptyBindings, cloned_empty)
+
+        with deepcopy(self.bindings) as cloned:
+            with self.bindings.clone() as cloned_explicitly:
+                self.assertEqual(self.bindings, cloned)
+                self.assertEqual(cloned, cloned_explicitly)
+
+        # should be after deletion of cloned to ensure that clone is real
+        self.assertEqual(self.bindings, self.bindings)
+        self.assertEqual(self.emptyBindings, self.emptyBindings)
+        self.assertNotEqual(self.emptyBindings, self.bindings)
+
+    def test_bindings_merge(self):
+        merged_empty = Bindings.merge(self.emptyBindings, self.emptyBindings)
+        self.assertEqual(self.emptyBindings, merged_empty)
+
+        merged_with_empty_lr = Bindings.merge(self.emptyBindings, self.bindings)
+        merged_with_empty_rl = Bindings.merge(self.bindings, self.emptyBindings)
+        self.assertEqual(merged_with_empty_rl, merged_with_empty_lr)
+        self.assertEqual(merged_with_empty_rl, self.bindings)
+
+        merged_self = Bindings.merge(self.bindings, self.bindings)
+        self.assertEqual(merged_self, self.bindings)
+
+    def test_bindings_is_empty(self):
+        self.assertTrue(self.emptyBindings.is_empty())
+        self.assertFalse(self.bindings.is_empty())
+
+    def test_bindings_resolve(self):
+        # todo: how to perfom negative tests?
```

Comment:
Minor but it is better using `TODO` in upper case, because it is more easy to search it in a codebase.

Reply:
Regarding negative tests. From the Python API perspective `Bindings.resolve` should return `None` when key is absent. To not repeat the code may be it would be simpler to handle it in `Bindings.resolve` and check for the null pointer. If there are more specific questions in your mind we can discuss it.